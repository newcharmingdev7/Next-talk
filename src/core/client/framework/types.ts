// TODO: (@cvle) Extract useful common types into its own package.
export { Omit, Overwrite, PropTypesOf } from "coral-ui/types";
export { DeepPartial } from "coral-common/types";

export type RelayEnumLiteral<T> = keyof T | "%future added value";
