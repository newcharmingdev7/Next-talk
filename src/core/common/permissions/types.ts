export type UserRole =
  | "%future added value"
  | "COMMENTER"
  | "MEMBER"
  | "STAFF"
  | "MODERATOR"
  | "ADMIN";

export interface User {
  id: string;
  role: UserRole;
  moderationScopes?: null | {
    scoped?: boolean;
    siteIDs?: string[];
  };
}

export type Moderator = User & {
  role: "MODERATOR";
};

export type LTEModerator = User & {
  role: "MODERATOR" | "STAFF" | "MEMBER" | "COMMENTER";
};

export const isModerator = (user: User): user is Moderator => {
  return user.role === "MODERATOR";
};

export const isLTEModerator = (user: User): user is LTEModerator => {
  return user.role !== "ADMIN";
};
