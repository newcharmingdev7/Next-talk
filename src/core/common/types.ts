export type Diff<T extends keyof any, U extends keyof any> = ({ [P in T]: P } &
  { [P in U]: never } & { [x: string]: never })[T];

export type Omit<U, K extends keyof U> = Pick<U, Diff<keyof U, K>>;

export type Overwrite<T, U> = Pick<T, Diff<keyof T, keyof U>> & U;

export type Sub<T, U> = Pick<T, Diff<keyof T, keyof U>>;

/**
 * Make all properties in T writeable
 */
export type Writeable<T> = { -readonly [P in keyof T]: T[P] };

export type Promiseable<T> = Promise<T> | T;
