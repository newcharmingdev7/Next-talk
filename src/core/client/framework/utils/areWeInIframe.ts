/**
 * Returns true if we are in an iframe.
 */
export default function areWeInIframe() {
  try {
    // eslint-disable-next-line no-restricted-globals
    return window.self !== window.top;
  } catch (e) {
    return true;
  }
}
