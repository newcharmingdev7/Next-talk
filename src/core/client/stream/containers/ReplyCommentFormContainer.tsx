import React, { Component } from "react";
import { graphql } from "react-relay";

import { withContext } from "talk-framework/lib/bootstrap";
import { BadUserInputError } from "talk-framework/lib/errors";
import { withFragmentContainer } from "talk-framework/lib/relay";
import { PymStorage } from "talk-framework/lib/storage";
import { PropTypesOf } from "talk-framework/types";
import { ReplyCommentFormContainer_asset as AssetData } from "talk-stream/__generated__/ReplyCommentFormContainer_asset.graphql";
import { ReplyCommentFormContainer_comment as CommentData } from "talk-stream/__generated__/ReplyCommentFormContainer_comment.graphql";

import ReplyCommentForm, {
  ReplyCommentFormProps,
} from "../components/ReplyCommentForm";
import { CreateCommentMutation, withCreateCommentMutation } from "../mutations";

interface InnerProps {
  createComment: CreateCommentMutation;
  pymSessionStorage: PymStorage;
  comment: CommentData;
  asset: AssetData;
  onClose?: () => void;
}

interface State {
  initialValues?: ReplyCommentFormProps["initialValues"];
  initialized: boolean;
}

export class ReplyCommentFormContainer extends Component<InnerProps, State> {
  public state: State = { initialized: false };
  private contextKey = `replyCommentFormBody-${this.props.comment.id}`;

  constructor(props: InnerProps) {
    super(props);
    this.init();
  }

  private async init() {
    const body = await this.props.pymSessionStorage.getItem(this.contextKey);
    if (body) {
      this.setState({
        initialValues: {
          body,
        },
      });
    }
    this.setState({
      initialized: true,
    });
  }

  private handleOnCancel = () => {
    this.props.pymSessionStorage.removeItem(this.contextKey);
    if (this.props.onClose) {
      this.props.onClose();
    }
  };

  private handleOnSubmit: ReplyCommentFormProps["onSubmit"] = async (
    input,
    form
  ) => {
    try {
      await this.props.createComment({
        assetID: this.props.asset.id,
        parentID: this.props.comment.id,
        ...input,
      });

      this.props.pymSessionStorage.removeItem(this.contextKey);
      if (this.props.onClose) {
        this.props.onClose();
      }
    } catch (error) {
      if (error instanceof BadUserInputError) {
        return error.invalidArgsLocalized;
      }
      // tslint:disable-next-line:no-console
      console.error(error);
    }
    return undefined;
  };

  private handleOnChange: ReplyCommentFormProps["onChange"] = state => {
    if (state.values.body) {
      this.props.pymSessionStorage.setItem(this.contextKey, state.values.body);
    } else {
      this.props.pymSessionStorage.removeItem(this.contextKey);
    }
  };

  public render() {
    if (!this.state.initialized) {
      return null;
    }
    return (
      <ReplyCommentForm
        id={this.props.comment.id}
        onSubmit={this.handleOnSubmit}
        onChange={this.handleOnChange}
        initialValues={this.state.initialValues}
        onCancel={this.handleOnCancel}
      />
    );
  }
}
const enhanced = withContext(({ pymSessionStorage }) => ({
  pymSessionStorage,
}))(
  withCreateCommentMutation(
    withFragmentContainer<InnerProps>({
      asset: graphql`
        fragment ReplyCommentFormContainer_asset on Asset {
          id
        }
      `,
      comment: graphql`
        fragment ReplyCommentFormContainer_comment on Comment {
          id
        }
      `,
    })(ReplyCommentFormContainer)
  )
);
export type PostCommentFormContainerProps = PropTypesOf<typeof enhanced>;
export default enhanced;
