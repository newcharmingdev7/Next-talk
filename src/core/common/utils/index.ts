export { default as timeout } from "./timeout";
export { default as pascalCase } from "./pascalCase";
