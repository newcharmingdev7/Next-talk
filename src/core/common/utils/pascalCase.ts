import { camelCase } from "lodash";

export default function pascalCase(value: string) {
  return value[0].toUpperCase() + camelCase(value).slice(1);
}
