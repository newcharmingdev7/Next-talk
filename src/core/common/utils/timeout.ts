/** A promisified timeout. */
export default function timeout(ms: number = 0) {
  return new Promise(resolve => setTimeout(resolve, ms));
}
