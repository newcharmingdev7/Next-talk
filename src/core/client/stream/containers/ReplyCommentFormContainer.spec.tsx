import { shallow } from "enzyme";
import { noop } from "lodash";
import React from "react";
import sinon from "sinon";

import { PropTypesOf } from "talk-framework/types";

import { timeout } from "talk-common/utils";
import {
  createFakePymStorage,
  removeFragmentRefs,
} from "talk-framework/testHelpers";
import { ReplyCommentFormContainer } from "./ReplyCommentFormContainer";

const ReplyCommentFormContainerN = removeFragmentRefs(
  ReplyCommentFormContainer
);

function getContextKey(commentID: string) {
  return `replyCommentFormBody-${commentID}`;
}

it("renders correctly", async () => {
  const props: PropTypesOf<typeof ReplyCommentFormContainerN> = {
    // tslint:disable-next-line:no-empty
    createComment: (() => {}) as any,
    asset: {
      id: "asset-id",
    },
    comment: {
      id: "comment-id",
    },
    pymSessionStorage: createFakePymStorage(),
    autofocus: false,
  };

  const wrapper = shallow(<ReplyCommentFormContainerN {...props} />);
  await timeout();
  wrapper.update();
  expect(wrapper).toMatchSnapshot();
});

it("renders with initialValues", async () => {
  const props: PropTypesOf<typeof ReplyCommentFormContainerN> = {
    // tslint:disable-next-line:no-empty
    createComment: (() => {}) as any,
    asset: {
      id: "asset-id",
    },
    comment: {
      id: "comment-id",
    },
    pymSessionStorage: createFakePymStorage(),
    autofocus: false,
  };

  await props.pymSessionStorage.setItem(
    getContextKey(props.comment.id),
    "Hello World!"
  );

  const wrapper = shallow(<ReplyCommentFormContainerN {...props} />);
  await timeout();
  wrapper.update();
  expect(wrapper).toMatchSnapshot();
});

it("save values", async () => {
  const props: PropTypesOf<typeof ReplyCommentFormContainerN> = {
    // tslint:disable-next-line:no-empty
    createComment: (() => {}) as any,
    asset: {
      id: "asset-id",
    },
    comment: {
      id: "comment-id",
    },
    pymSessionStorage: createFakePymStorage(),
    autofocus: false,
  };

  await props.pymSessionStorage.setItem(
    getContextKey(props.comment.id),
    "Hello World!"
  );

  const wrapper = shallow(<ReplyCommentFormContainerN {...props} />);
  await timeout();
  wrapper.update();
  wrapper
    .first()
    .props()
    .onChange({ values: { body: "changed" } });
  expect(
    await props.pymSessionStorage.getItem(getContextKey(props.comment.id))
  ).toBe("changed");
});

it("creates a comment", async () => {
  const assetID = "asset-id";
  const input = { body: "Hello World!" };
  const createCommentStub = sinon.stub();
  const form = { reset: noop };
  const onCloseStub = sinon.stub();

  const props: PropTypesOf<typeof ReplyCommentFormContainerN> = {
    // tslint:disable-next-line:no-empty
    createComment: createCommentStub,
    asset: {
      id: "asset-id",
    },
    comment: {
      id: "comment-id",
    },
    pymSessionStorage: createFakePymStorage(),
    onClose: onCloseStub,
    autofocus: false,
  };

  await props.pymSessionStorage.setItem(
    getContextKey(props.comment.id),
    "Hello World!"
  );

  const wrapper = shallow(<ReplyCommentFormContainerN {...props} />);
  await timeout();
  wrapper.update();
  wrapper
    .first()
    .props()
    .onSubmit(input, form);
  expect(
    createCommentStub.calledWith({
      assetID,
      parentID: props.comment.id,
      ...input,
    })
  ).toBeTruthy();
  await timeout();
  expect(onCloseStub.calledOnce).toBe(true);
});

it("closes on cancel", async () => {
  const onCloseStub = sinon.stub();
  const props: PropTypesOf<typeof ReplyCommentFormContainerN> = {
    // tslint:disable-next-line:no-empty
    createComment: (() => {}) as any,
    asset: {
      id: "asset-id",
    },
    comment: {
      id: "comment-id",
    },
    pymSessionStorage: createFakePymStorage(),
    onClose: onCloseStub,
    autofocus: false,
  };

  await props.pymSessionStorage.setItem(
    getContextKey(props.comment.id),
    "Hello World!"
  );

  const wrapper = shallow(<ReplyCommentFormContainerN {...props} />);
  await timeout();
  wrapper.update();
  wrapper.findWhere(w => !!w.prop("onCancel")).prop("onCancel")();

  // Calls close.
  expect(onCloseStub.calledOnce).toBe(true);

  // Removes saved value.
  expect(
    await props.pymSessionStorage.getItem(getContextKey(props.comment.id))
  ).toBeNull();
});
