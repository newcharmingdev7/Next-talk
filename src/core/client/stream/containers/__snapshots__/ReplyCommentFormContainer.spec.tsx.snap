// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<ReplyCommentForm
  onCancel={[Function]}
  onChange={[Function]}
  onSubmit={[Function]}
/>
`;

exports[`renders with initialValues 1`] = `
<ReplyCommentForm
  initialValues={
    Object {
      "body": "Hello World!",
    }
  }
  onCancel={[Function]}
  onChange={[Function]}
  onSubmit={[Function]}
/>
`;
